fem/normal_deriv_restriction.cpp

```diff
@@ -0,0 +1,854 @@
+// Copyright (c) 2010-2024, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "normal_deriv_restriction.hpp"
+#include "fespace.hpp"
+#include "pgridfunc.hpp"
+#include "fe/face_map_utils.hpp"
+#include "../general/forall.hpp"
+
+namespace mfem
+{
+
+/// Compute the face index to volume index map "face_to_vol" in 2D
+static void NormalDerivativeSetupFaceIndexMap2D(
+   int nf, int d, const Array<int>& face_to_elem, Array<int>& face_to_vol)
+{
+   const auto f2e = Reshape(face_to_elem.HostRead(), 2, 2, nf);
+   auto f2v = Reshape(face_to_vol.HostWrite(), d, 2, nf);
+
+   for (int f = 0; f < nf; ++f)
+   {
+      const int fid0 = f2e(0, 1, f);
+      const int fid1 = f2e(1, 1, f);
+      for (int side = 0; side < 2; ++side)
+      {
+         const int el = f2e(side, 0, f);
+
+         if (el < 0)
+         {
+            for (int p = 0; p < d; ++p)
+            {
+               f2v(p, side, f) = -1;
+            }
+         }
+         else
+         {
+            for (int p = 0; p < d; ++p)
+            {
+               int i, j;
+               internal::FaceIdxToVolIdx2D(p, d, fid0, fid1, side, i, j);
+
+               f2v(p, side, f) = i + d * j;
+            }
+         }
+      }
+   }
+}
+
+/// Compute the face index to volume index map "face_to_vol" in 3D
+static void NormalDerivativeSetupFaceIndexMap3D(
+   int nf, int d, const Array<int>& face_to_elem, Array<int>& face_to_vol)
+{
+   const auto f2e = Reshape(face_to_elem.HostRead(), 2, 3, nf);
+   auto f2v = Reshape(face_to_vol.HostWrite(), d*d, 2, nf);
+
+   for (int f = 0; f < nf; ++f)
+   {
+      const int fid0 = f2e(0, 1, f);
+      const int fid1 = f2e(1, 1, f);
+      for (int side = 0; side < 2; ++side)
+      {
+         const int el = f2e(side, 0, f);
+         const int orientation = f2e(side, 2, f);
+
+         if (el < 0)
+         {
+            for (int p = 0; p < d*d; ++p)
+            {
+               f2v(p, side, f) = -1;
+            }
+         }
+         else
+         {
+            for (int p = 0; p < d*d; ++p)
+            {
+               int i, j, k; // 3D lexicographic index of quad point p
+               internal::FaceIdxToVolIdx3D(p, d, fid0, fid1, side, orientation, i, j, k);
+
+               f2v(p, side, f) = i + d * (j + d * k);
+            }
+         }
+      }
+   }
+}
+
+L2NormalDerivativeFaceRestriction::L2NormalDerivativeFaceRestriction(
+   const FiniteElementSpace &fes_,
+   const ElementDofOrdering ordering,
```

Comment:
For consistency with the declaration:
```suggestion
   const ElementDofOrdering f_ordering,
```

Reply:
Currently, this argument is not used. The assumption is that it is `ElementDofOrdering::LEXICOGRAPHIC`, as in `L2FaceRestriction::CheckFESpace()`. For clarity, it is good to add a comment about this and add an `MFEM_ASSERT`/`MFEM_VERIFY` that the other case is not implemented.

Reply:
Done in fd28c2809.